impeller/docs/babys_first_triangle.md

```diff
@@ -0,0 +1,276 @@
+
+Shaders define the programmable stages of a pipeline. We are going to be define a vertex and fragment shader for our triangle.
+
+The job of a vertex shader is to transform the vertices of our triangle into [normalized device coordinates](coordinate_system.md) (NDC). The rasterizer will then take these coordinates and convert them to 2D coordinates in the framebuffer.
+
+OTOH, the job of the fragment shader is to shade (color) the pixels covered by the triangle.
+
+#### Vertex Shader
+
+Let's first create `baby.vert`:
+
+```glsl
+in vec2 position;
+
+void main() {
+  gl_Position = vec4(position, 0.0, 1.0);
+}
+```
+
+This shader, which expects to run once per vertex, takes a `vec2` and converts into into NDC. Are you can see, there is no "conversion" going on. That's because our vertices will already be in NDC. Since we are only drawing a simple triangle, we will need to give it three vertices. We'll discuss that in later setup.
+
+#### Fragment Shader
+
+And, `baby.frag`:
+
+```glsl
+out vec4 frag_color;
+
+void main() {
+  frag_color = vec4(1.0, 0.0, 0.0, 1.0);
+}
+
+```
+
+This shader, which expects to run once per texture element (texel) covered by the triangle just returns a solid red color as output.
+
+### The Pipeline Descriptor
+
+Invoking the shader compiler will generate the backend specific shaders (GLSL ES for OpenGL ES, Metal Shading Language code for Metal, and SPIRV for Vulkan). Along with these artifacts, the compiler will generate a couple of C++ header files that contain the interfaces and metadata you will need to create a pipeline with these shaders at runtime. Find those somewhere in the generated artifacts. These will be called `baby.vert.h` and `baby.frag.h` for each of our two shaders. Include then in your translation unit. We need these for the pipeline descriptor.
+
+But first, let's take a peek inside `baby.vert.h` to look at what the compiler gave us. There should be a whole bunch of metadata you don't really need to care about. But there is one struct called `PerVertexData` that looks interesting:
+
+```c++
+struct PerVertexData {
+  Point position;  // (offset 0, size 8)
+};  // struct PerVertexData (size 8)
+```
+
+The compiler has detected that the shader expects one point position per vertex. It is going to be our job to fill this in during rendering.
+
+This struct is handy because as you tinker on your shader, the compiler will add, remove, and reorder the fields. If there are alignment consideration for the GPU, the compiler knows about these and it will add the appropriate padding between these fields so you all you have to worry about is filling in the position. You don't have to use this struct directly, but trusting the compiler will greatly simplify you experience.
```

Comment:
```suggestion
This struct is handy because as you tinker on your shader, the compiler will add, remove, and reorder the fields. If there are alignment considerations for the GPU, the compiler knows about these and it will add the appropriate padding between these fields so you all you have to worry about is filling in the position. You don't have to use this struct directly, but trusting the compiler will greatly simplify you experience.
```